14.0 mig product_contract

mourad-ehm opened the pull request.

Commit:
[ADD] product_contract: Create module
* Add contract functionality to `product.templates`
* Add logic to create contracts from `sale.order` that contains contract products.

oca-transbot committed:
OCA Transbot updated translations from Transifex

lasley committed:
[FIX] product_contract: Fix mock usage in tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

flotho committed:
[FIX] contract_sale_generation: it doesn't create sales (#141)

* Change the method called in the view
* Complete the create_invoice method
* Bump version + authoring
* Correct bad call of method
  Small Documentation
* Add super call in python test
* FIX bad field names causing bad quantities in sale.order.line

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-travis committed:
[UPD] Update product_contract.pot

sbejaoui committed:
[MIG] - Product Contract Migration to 12.0

sbejaoui committed:
[IMP] - Product with is_contract can be only of type service

sbejaoui committed:
[IMP] - Change dependencies to contract_sale

sbejaoui committed:
[IMP] - Add recurrence fields to product template and sale order line

sbejaoui committed:
[IMP] - Create contract on sale order confirmation

- On Sale Order confirmation, a contract is created for each contract template used on sale order lines
- A not finished contract can be mentioned on sale order line
- A sale order line linked to a contract will update it and don't create a new one if it had the same template

sbejaoui committed:
[IMP] - Link contract line to sale order line

sbejaoui committed:
[FIX] - Remove recurring_next_date from sale order line

recurring_next_date should be computed by contract line to get default value

sbejaoui committed:
[IMP] - Contract product are ignored on invoicing process

- Sale order line for contract product pass to nothing to invoice on order confirmation
- Contract Invoices are linked to sale order line

sbejaoui committed:
[FIX] - Default value for date_start applied on product change

sbejaoui committed:
[ADD] - Add readme directory

sbejaoui committed:
[IMP] - Onchange contract product on contract contract and contract template

sbejaoui committed:
[FIX] - Change website to OCA repository url and prefix module name with Recurring

sbejaoui committed:
[FIX] - Compute recurring_next_date before create contract line

sbejaoui committed:
[IMP] - Upsell/Downsell contract from sale order

sbejaoui committed:
[ADD] - Add renewal process with termination notice

sbejaoui committed:
[IMP] - compute date end onchange date start for auto-renew sale order lines

sbejaoui committed:
[IMP] - link contract line and its successor in upsel case

sbejaoui committed:
[FIX] - on upsel, contract line should stop day - 1

sbejaoui committed:
[FIX] - fix onchange

sbejaoui committed:
[IMP] - hide recurring_invoicing_type if recurring_rule_type is monthlylastday

for the monthlylastday case, pre-paid is logicly impossible,
if monthlylastday is set, we consider only post-paid case

sbejaoui committed:
[FIX] - include date_end in the period if the product is_autorenew

sbejaoui committed:
[IMP] - _prepare_contract_value for sale order confirm

sbejaoui committed:
[REM] - Remove unused method

sbejaoui committed:
[IMP] - get contract user from sale order user

sbejaoui committed:
[IMP] - show related sale orders in contract form

ThomasBinsfeld committed:
[REF] Contract Product: invoice in prepare_invoice_line is optional

sbejaoui committed:
[FIX] - fix flake8

ThomasBinsfeld committed:
[ADD] Product Contract: payment term

sbejaoui committed:
[IMP] - update invoice_line vals only if it is not null

sbejaoui committed:
[FIX] - fix unit tests

sbejaoui committed:
[IMP] - Add unit test

sbejaoui committed:
[REF] - predecessor_contract_line set in create process

sbejaoui committed:
[IMP] - Simplify sale order line creation for contract product

ThomasBinsfeld committed:
[ADD] Contract Sale: fiscal position

sbejaoui committed:
[IMP] - auto_renew_rule deducted from recurring_rule_type and product_uom_qty

ThomasBinsfeld committed:
[FIX] Product Contract: copy invoicing partner to created contract

sbejaoui committed:
[FIX] - Fix recurring_next_date

sbejaoui committed:
[FIX] - date_end required only for contract product

sbejaoui committed:
[FIX] - Get auto_renew_rule_type based on recurring_rule_type

sbejaoui committed:
[IMP] - cancel contract line if the downsell/upsell comes before

sbejaoui committed:
[FIX] - date_end computed only for contract products

ThomasBinsfeld committed:
[REF] Contract: set order count to 0 if no access to orders

sbejaoui committed:
[IMP] - permit contract selection without contract_template_id

qgroulard committed:
[ADD] Add start_date to contract_line display_name

In order to help distinguishing lines with the same name

sbejaoui committed:
[12.0][IMP] - option to decide if we automaticallt create contract a sale confirm

sbejaoui committed:
[FIX] - fix display name compute dependecy

sbejaoui committed:
[12.0][FIX] - create contract only on sale and done states

sbejaoui committed:
[12.0][REF] - Split analytic accouts and contracts

sbejaoui committed:
[12.0][IMP] - Improve unit tests

sbejaoui committed:
[IMP] - improve is contract visibility in product view

sbejaoui committed:
[12.0][IMP] - sale line linked to contracts must not be invoiced from sale order

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

sbejaoui committed:
[FIX] - Remove deprecated field

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

OCA-git-bot committed:
product_contract 12.0.2.0.1

marylla committed:
Translated using Weblate (German)

Currently translated at 61.0% (25 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/de/

britoederr committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

sbejaoui committed:
[12.0][FIX] - contract company must be the same as the sale order

When creating a contract from a sale order the company must be the sale order
company and not the user company

rschnapka committed:
Translated using Weblate (German)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/de/

OCA-git-bot committed:
product_contract 12.0.2.0.2

sbejaoui committed:
[12.0][FIX] - Run onchange contract_type when changing the contract template

Fix this use-case:

If the contract journal is not set on the contract template the contract is created
without a journal when confirming the sale order

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/hr/

Martronic-SA committed:
Translated using Weblate (French)

Currently translated at 92.7% (38 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/fr/

sbidoul committed:
[IMP] contract: support pre-paid for monthlylastday

monthlylastday is (almost) not a special case anymore \o/.
montlylastday is simply a montly period where the
periods are aligned on month boundaries.
The last bit of special casing is that postpaid generates
invoice the day after the last dasy of the period, except
for monthlylastday where the invoice is generated on the
last day of the period. This last exception will disappear
when we put the offset under user control.

This is a breaking change because the post-paid/pre-paid
mode becomes relevant for monthlylastday invoicing.
The field becomes visible in the UI. Code that generate
monthlylastday contract lines must now correctly set
the pre-paid/post-paid mode too. Some tests have had
to be adapted to reflect that.

sbejaoui committed:
[IMP] - Set recurring_invoicing_type for existing products set to monthlylastday

OCA-git-bot committed:
product_contract 12.0.3.0.0

sbejaoui committed:
[12.0][FIX] - Set contract line analytic account at contract creation from SO

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
product_contract 12.0.3.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

pedrobaeza committed:
[IMP] contract: remove redundant method

I have detected a method that was created as redundant and with the same
technique used when preparing the line values, so better to have
everything together in the same method instead of having it spread.

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/es/

OCA-git-bot committed:
product_contract 12.0.3.1.0

OCA-git-bot committed:
product_contract 12.0.3.1.1

sbejaoui committed:
[IMP] - can't upsell or downsell a resiliated contract

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/it/

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/es/

Commit:
[FIX]contract_count == len(active contracts)

Commit:
[FIX] product_contract: set 'Contract template' field company depend

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 73.2% (30 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt/

sbejaoui committed:
[12.0][ADD] - add auto_renew_interval, auto_renew_rule_type fields to sale order line

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 73.2% (30 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt/

alvarorib committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (41 of 41 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt/

sbejaoui committed:
[IMP] - Improve UX, open contract form if the sale order is related to one (same for contract)

Commit:
[FIX] contract_sale_mandate: use property_contract_template_id

OCA-git-bot committed:
product_contract 12.0.3.1.2

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

sbejaoui committed:
[IMP] - use terminate term instead of resiliate

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
product_contract 12.0.5.0.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

svalaeys committed:
Translated using Weblate (French)

Currently translated at 100.0% (43 of 43 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/fr/

sbejaoui committed:
[12.0][IMP] - Add failing test for creating contract grouped by contract template

When creating a contract from SO, the systme should bgoup by contract template
sale order lines and create one contract per contract template

actually the system group by contract template but create as many contracts as many
sale order lines.

sbejaoui committed:
[12.0][FIX] - fix contract creation group by contract template

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
product_contract 12.0.5.0.1

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

OCA-git-bot committed:
product_contract 12.0.5.0.2

sergiocorato committed:
[FIX] migration from 11.0

Update product_contract/migrations/12.0.3.0.0/pre-migration.py

Co-Authored-By: Bejaoui Souheil <souheil_bejaoui@hotmail.fr>

Update product_contract/migrations/12.0.3.0.0/pre-migration.py

Co-Authored-By: Bejaoui Souheil <souheil_bejaoui@hotmail.fr>

[FIX]

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (47 of 47 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/es/

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
product_contract 12.0.5.0.3

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
product_contract 12.0.5.0.4

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

sbejaoui committed:
[IMP] - Contract: add new recurring_rule_type quarterly and semesterly

eduaparicio committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (47 of 47 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

oca-travis committed:
[UPD] Update product_contract.pot

OCA-git-bot committed:
product_contract 12.0.5.1.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

oca-travis committed:
[UPD] Update product_contract.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

fcolus committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (49 of 49 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

pedrobaeza committed:
Translated using Weblate (Spanish)

Currently translated at 95.9% (47 of 49 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/es/

fcolus committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (49 of 49 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

fcolus committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (49 of 49 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/pt_BR/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/

gregory-moka committed:
Translated using Weblate (French)

Currently translated at 100.0% (49 of 49 strings)

Translation: contract-12.0/contract-12.0-product_contract
Translate-URL: https://translation.odoo-community.org/projects/contract-12-0/contract-12-0-product_contract/fr/

sbejaoui committed:
[13.0][MIG] - migration product_contract

Commit:
[REF] product_contract: micro-refactoring to ease overrides

mourad-ehm committed:
[MIG] product_contract: Migration to 14.0

hparfr commented:
@mourad-ehm please fix pre-commit and remove .format() as requested in v13 port PR.

oa-eska commented:
https://github.com/OCA/contract/pull/590 this can be added as well.

hparfr commented:
@mourad-ehm can you update this pr ?

lukebranch commented:
@mourad-ehm ,

Please let me know if you would like any additional review or if you would be open to PR's against akretion/contract/tree/14.0-mig-product_contract. We have a need for this module in 14.0 and would like to offer an extra pair of hands if you open to this.

lukebranch commented:
> @mourad-ehm ,
> 
> Please let me know if you would like any additional review or if you would be open to PR's against akretion/contract/tree/14.0-mig-product_contract. We have a need for this module in 14.0 and would like to offer an extra pair of hands if you open to this.

@mourad-ehm ,

Not sure if you're still planning on working on this migration, but please let me know if you would like someone to take over and work on making the final changes required to take your product_contract migration over the finish line so it's ready to be merged.

We'd like to make use of this feature in 14.0 for a project but would like to make sure it's in stable 14.0 OCA/contract before making use of it.

hparfr commented:
@lukebranch you can take over this migration.
I don't think there will be any activity on this PR in the upcoming weeks.

lukebranch commented:
> @lukebranch you can take over this migration.
> I don't think there will be any activity on this PR in the upcoming weeks.

@hparfr,

Apologies for the late reply. Thanks for letting me know. 

@mourad-ehm, please let me know if you end up wanting to pick this back up.

@hparfr, as this would be our first OCA contribution, I may need some time to get up to speed on OCA contribution requirements and processes. My apologies in advance for this.

In regards to picking up on the already substantial work that @mourad-ehm has already done, what would be the best way for me to get started? Should I fork the original repo at https://github.com/akretion/contract/tree/14.0-mig-product_contract and submit a new PR?

designbybeck commented:
@mourad-ehm @lukebranch  We are needing this in v14 as well. Anything we can do to help? We had some discussions about this on the Odoo OCA Matrix/Element chatroom (#oca-general:matrix.org). Should we just have those folks reply here (GitHub) for discussion on trying to get Product_Contracts working in v14?

@hparfr Why do you say there will be no work on it in the next few weeks?

Thank you.

hparfr commented:
> @hparfr Why do you say there will be no work on it in the next few weeks?

Mostly because of holidays.

> In regards to picking up on the already substantial work that @mourad-ehm has already done, what would be the best way for me to get started? Should I fork the original repo at https://github.com/akretion/contract/tree/14.0-mig-product_contract and submit a new PR?

* This PR has been started a while ago, current v12 (and maybe v13) version have been improved. 
* There is no new code in this PR (apart version in __manifest__ )
 
So, I think it's better to re-start a migration and create your own PR.

Do not hesitate to read the migration guide : https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-14.0

centrix-luke commented:
@designbybeck @lukebranch @mourad-ehm @hparfr ,

We're also interested in migrating this module. We don't have the manpower on hand currently due to project workload, however we'd be interested to discuss if everyone wanted to pool resources to get this migrated to 14.0.

Is there an OCA-approved way to contribute resources towards a module migration, similar to a bug bounty?

mourad-ehm committed:
[14.0][FIX] product_contract: pre-commit format code

mourad-ehm committed:
[14.0][FIX] product_contract: use with_company

rousseldenis commented:
@mourad-ehm Could you please enhance your commit messages (something more like '[14.0][FIX] product_contract: Use with_company')  ?

rousseldenis commented:
@hparfr  is this really on stale on your side ?

centrix-luke commented:
> @designbybeck @lukebranch @mourad-ehm @hparfr ,
> 
> We're also interested in migrating this module. We don't have the manpower on hand currently due to project workload, however we'd be interested to discuss if everyone wanted to pool resources to get this migrated to 14.0.
> 
> Is there an OCA-approved way to contribute resources towards a module migration, similar to a bug bounty?

Further to this comment a few months ago, we're interested in organising to pay someone else to migrate product_contract and build a solution for allowing e-commerce sales for subscription/payment by instalment (eg. select weekly, fortnightly, monthly, yearly variant on the product page in e-commerce to pay for the product in instalments, or perhaps purchase a yearly subscription). 

I'm not sure if it's appropriate to write this here, so please advise if inappropriate and i'll delete the comment, however I feel this perhaps requires funding in order to push it forwards for version 14.0, which we are willing to discuss. Anyone interested in discussing please reach out to us on info@centrix.io. All code would be contributed back to the OCA and/or placed in a public repository.

hparfr commented:
@mourad-ehm can you please integrate #590 and amend your commit as Denis suggested ?

mourad-ehm committed:
[14.0][FIX] product_contract: pre-commit

[14.0][FIX] product_contract: action_show_contracts rights

[14.0][FIX] product_contract: replace with_context force_company by with_company

[FIX][14.0] product_contract: pre-commit

[FIX][14.0] product_contract: add setup/product_contract

mourad-ehm commented:
Hi @rousseldenis, @hparfr I enhanced commit messages and I integrated commit of pr  https://github.com/OCA/contract/pull/590

rousseldenis reviewed (approved):
@mourad-ehm Thanks.

I encourage you to install pre-commit locally before committing in order to avoid separate pre-commit fixes (it does it automatically).

mourad-ehm commented:
Thanks @rousseldenis for review

hparfr reviewed (approved):
LGTM

scigghia commented:
Hello,
Would someone merge this PR ?

hparfr commented on product_contract/models/sale_order.py:
```diff
@@ -127,7 +127,7 @@ def _compute_contract_count(self):
     def action_show_contracts(self):
         self.ensure_one()
-        action = self.env.ref("contract.action_customer_contract").read()[0]
+        action = self.env.ref("contract.action_customer_contract").sudo().read()[0]
```
```suggestion
        action = self.env["ir.actions.act_window"]._for_xml_id("contract.action_customer_contract")
```

mourad-ehm replied:
Without sudo ?

mourad-ehm replied:
@hparfr

yweng8111 commented on product_contract/models/sale_order.py:
```diff
@@ -0,0 +1,146 @@
+            contract_templates = self.env["contract.template"]
+            for order_line in line_to_create_contract:
+                contract_template = order_line.product_id.with_context(
+                    force_company=rec.company_id.id
```
Please use "with_company()" instead of "with_context(force_company=xxx)"

yweng8111 commented on product_contract/models/sale_order.py:
```diff
@@ -0,0 +1,146 @@
+            for contract_template in contract_templates:
+                order_lines = line_to_create_contract.filtered(
+                    lambda r, template=contract_template: r.product_id.with_context(
+                        force_company=r.order_id.company_id.id
```
Please use "with_company()" instead of "with_context(force_company=xxx)"

yweng8111 commented on product_contract/models/product_template.py:
```diff
@@ -0,0 +1,79 @@
+    def write(self, vals):
+        if "is_contract" in vals and vals["is_contract"] is False:
+            for company in self.env["res.company"].search([]):
+                self.with_context(force_company=company.id).write(
```
Please use "with_company()" instead of "with_context(force_company=xxx)"

yweng8111 commented on product_contract/tests/test_sale_order.py:
```diff
@@ -0,0 +1,362 @@
+                ],
+            }
+        )
+        self.product1.with_context(force_company=self.sale.company_id.id).write(
```
Please use with_company(self.sale.company_id) instead of with_context(force_company=self.company_id.id)

yweng8111 commented on product_contract/tests/test_sale_order.py:
```diff
@@ -0,0 +1,362 @@
+                "property_contract_template_id": self.contract_template1.id,
+            }
+        )
+        self.product2.with_context(force_company=self.sale.company_id.id).write(
```
Please use with_company(self.sale.company_id) instead of with_context(force_company=self.company_id.id)

jesusVMayor commented on product_contract/views/sale_order.xml:
```diff
@@ -1,102 +1,124 @@
-                        string="Create Contracts"
-                        type="object"
-                        class="oe_highlight" attrs="{'invisible': [('need_contract_creation', '=', False)]}"/>
+                <field name="need_contract_creation" attrs="{'invisible': []}" />
```
this field could be always invisible

jesusVMayor commented on product_contract/views/sale_order.xml:
```diff
@@ -0,0 +1,124 @@
+                </button>
+            </xpath>
+            <xpath expr="//field[@name='order_line']" position="before">
+                <field name="is_contract" attrs="{'invisible': []}" />
```
like need_contract_creation could be invisible="1"

jesusVMayor commented on product_contract/views/sale_order.xml:
```diff
@@ -0,0 +1,124 @@
+                    attrs="{'column_invisible': [('parent.is_contract', '=', False)]}"
+                />
+            </xpath>
+            <xpath expr="//field[@name='order_line']/tree" position="attributes">
```
This makes the module incompatible with  sale_product_configurator module, but i don't see a good way to make them work together. 
Maybe adding a button into the tree view to open the contract configuration fields of the line?

scigghia replied:
I confirm what @jesusVMayor says
Maybe adding a tab in the sale configurator popup with contract configurator fields, only when sale_product_configurator is installed, otherwise it is ok as it is

rousseldenis commented:
/ocabot migration product_contract

rousseldenis commented:
@mourad-ehm Could you check pre-commit and attend last comments?

elvise commented:
@mourad-ehm can you check pre-commit ?

elvise commented:
@mourad-ehm can you fix pre-commit and test ?

mourad-ehm commented:
> @mourad-ehm can you fix pre-commit and test ?

I did it but the error persist. I fixed it by this commande in local 'pre-commit run --show-diff-on-failure --color=always --all-files' Do you have any suggestion ?

rousseldenis commented:
> > @mourad-ehm can you fix pre-commit and test ?
> 
> I did it but the error persist. I fixed it by this commande in local 'pre-commit run --show-diff-on-failure --color=always --all-files' Do you have any suggestion ?

@mourad-ehm Could you try to rebase locally?

mourad-ehm commented:
I rebased branch locally before fix pre-commit

rousseldenis commented:
@mourad-ehm Do a git status locally. You'll see unadded setup folder.

Tested with your branch:

![image](https://user-images.githubusercontent.com/19529533/184873662-ad27b1c9-06d4-48f3-8524-61db5964f552.png)

mourad-ehm commented:
Thanks @rousseldenis

elvise commented:
> Thanks @rousseldenis 

Hi @mourad-ehm Thanks for rebase! 
Can you fix the test ?

rousseldenis commented:
@mourad-ehm Could you squash not interesting commits ?

rousseldenis commented:
> > Thanks @rousseldenis
> 
> Hi @mourad-ehm Thanks for rebase! Can you fix the test ?

Which test ?

elvise commented:
> > > Thanks @rousseldenis
> > 
> > 
> > Hi @mourad-ehm Thanks for rebase! Can you fix the test ?
> 
> Which test ?
@rousseldenis 

![IMG_1237](https://user-images.githubusercontent.com/5269577/185092764-8c64c0ef-c19e-461b-a2cb-dd137e183999.jpeg)

rousseldenis commented:
> > > > Thanks @rousseldenis
> > > 
> > > 
> > > Hi @mourad-ehm Thanks for rebase! Can you fix the test ?
> > 
> > 
> > Which test ?
> > @rousseldenis
> 
> ![IMG_1237](https://user-images.githubusercontent.com/5269577/185092764-8c64c0ef-c19e-461b-a2cb-dd137e183999.jpeg)

Code coverage is not mandatory

elvise commented:
> > > > > Thanks @rousseldenis
> > > > 
> > > > 
> > > > Hi @mourad-ehm Thanks for rebase! Can you fix the test ?
> > > 
> > > 
> > > Which test ?
> > > @rousseldenis
> > 
> > 
> > ![IMG_1237](https://user-images.githubusercontent.com/5269577/185092764-8c64c0ef-c19e-461b-a2cb-dd137e183999.jpeg)
> 
> Code coverage is not mandatory

Just because it failing 🙂

elvise commented:
Hi @mourad-ehm can you hide these fields ?


![IMG_1238](https://user-images.githubusercontent.com/5269577/185122711-fcb5b87e-d7ba-4ecd-a07f-b7fdb10fe8aa.jpeg)

elvise commented:
@mourad-ehm from my perspective the user needs to select also the “Recurring Type” from the product form

![IMG_1239](https://user-images.githubusercontent.com/5269577/185125687-d8435768-c4d0-4e8a-b9b5-1aa8fe92580a.jpeg)


![IMG_1240](https://user-images.githubusercontent.com/5269577/185125947-819836f0-d0e1-4143-aba1-9533b37f1c21.jpeg)


What do you think ?

mourad-ehm commented:
Hi @elvise, I think that we can add several improvement to this module. The best way is to merge first this PR than propose new PRs for improvements. The PR was created most than 1 year and the V 16 come soon. Improvements can be more pertinent in the new version.

elvise commented:
> Hi @elvise, I think that we can add several improvement to this module. The best way is to merge first this PR than propose new PRs for improvements. The PR was created most than 1 year and the V 16 come soon. Improvements can be more pertinent in the new version. 

@mourad-ehm from my perspective we need to merge a clear form, and frankly speaking i think we need to polish it. 
But anyway this is only my opinion 😅

pere-aquarian commented on product_contract/models/contract.py:
```diff
@@ -0,0 +1,38 @@
+    @api.depends("contract_line_ids")
+    def _compute_sale_order_count(self):
+        for rec in self:
+            try:
```
Seems that there are no records in self, so maybe there must be a sale order related to the contrat at test...

pere-aquarian commented on product_contract/models/contract.py:
```diff
@@ -0,0 +1,38 @@
+                order_count = len(
+                    rec.contract_line_ids.mapped("sale_order_line_id.order_id")
+                )
+            except AccessError:
```
Seem to be the same problem in the last comment.

pere-aquarian commented on product_contract/models/contract.py:
```diff
@@ -0,0 +1,38 @@
+
+    def action_view_sales_orders(self):
+        self.ensure_one()
+        orders = self.contract_line_ids.mapped("sale_order_line_id.order_id")
```
... and if we had is a sale_order created, orders should be not null.

pere-aquarian commented:
Hi @mourad-ehm and everybody!

Is there something I can do to help? 

I have reviewed the codecov warnings and seems that if we have a sale_order related to a contract some errors will disapear. Please: let me know if is there something else I can do. 

Thanks a lot!

francesco-ooops commented:
@OCA/project-service-maintainers @rousseldenis @pedrobaeza I see many approvals, what do we need for merge? 

thanks!

rousseldenis commented:
@mourad-ehm Could you check the flow in runboat as I have this when confirming sale order:

![image](https://user-images.githubusercontent.com/19529533/200273660-909c04d9-02d7-4485-a77a-7ebfe5a968e8.png)

francesco-ooops commented:
@mourad-ehm think you have some time to work on this soon? :)

francesco-ooops commented:
> @mourad-ehm Could you check the flow in runboat as I have this when confirming sale order:
> 
> ![image](https://user-images.githubusercontent.com/19529533/200273660-909c04d9-02d7-4485-a77a-7ebfe5a968e8.png)

@mourad-ehm sorry to push, do you think you can check this or shall we make a PR to supersede this one?

thanks

mourad-ehm commented:
Do you have some idea of origin of this issue ?

francesco-ooops commented:
@rousseldenis

rousseldenis commented:
No, I've done simple test, adding a contract product in a sale order and try to confirm.

francesco-ooops commented:
@mourad-ehm think you can find some time to fix the behavior this week?

francesco-ooops commented:
@rousseldenis could you make a video of the behavior? I could not reproduce it

rousseldenis commented:
@francesco-ooops @mourad-ehm 
Done on runboat:
[Capture vidéo du 24-11-22 09:21:17.webm](https://user-images.githubusercontent.com/19529533/203730233-0b202ba3-2a47-4f4f-98bf-87c9ca86651d.webm)

rousseldenis commented:
It seems that my capture does not like 'Create' :smiley: 

So:

- Just create a sale order, add the product 'Test Contract'.
- Save
- Confirm Sale
- Error

ilyasProgrammer commented:
> It seems that my capture does not like 'Create' 😃
> 
> So:
> 
>     * Just create a sale order, add the product 'Test Contract'.
> 
>     * Save
> 
>     * Confirm Sale
> 
>     * Error

@rousseldenis I cant repeat error after rebuild. Can you please check what is missing to reproduce error ?

rousseldenis commented:
> > It seems that my capture does not like 'Create' smiley
> > So:
> > ```
> > * Just create a sale order, add the product 'Test Contract'.
> > 
> > * Save
> > 
> > * Confirm Sale
> > 
> > * Error
> > ```
> 
> @rousseldenis I cant repeat error after rebuild. Can you please check what is missing to reproduce error ?

Ok, it's working.

rousseldenis commented:
/ocabot merge patch

rousseldenis commented:
@mourad-ehm Could you fix:

![image](https://user-images.githubusercontent.com/19529533/204496425-257ca202-df1a-4c1f-8e14-b87b4f7ff25d.png)


And could you check this : 

![image](https://user-images.githubusercontent.com/19529533/204496860-5d05f1b5-44b3-4e11-87e0-bb99915446b4.png)


The last build on 14.0 (5 days ago) suceeded

rousseldenis commented:
/ocabot rebase

francesco-ooops commented:
> @mourad-ehm Could you fix:
> 
> ![image](https://user-images.githubusercontent.com/19529533/204496425-257ca202-df1a-4c1f-8e14-b87b4f7ff25d.png)
> 
> And could you check this :
> 
> ![image](https://user-images.githubusercontent.com/19529533/204496860-5d05f1b5-44b3-4e11-87e0-bb99915446b4.png)
> 
> The last build on 14.0 (5 days ago) suceeded

@mourad-ehm think you have the time to check this? just for planning :)

aleixq commented:
I found the same issue as in: 

> @mourad-ehm Could you check the flow in runboat as I have this when confirming sale order:
> 
> ![image](https://user-images.githubusercontent.com/19529533/200273660-909c04d9-02d7-4485-a77a-7ebfe5a968e8.png)

I am trying to list all the steps I've done manually, trying to replicate the test in code /product_contract/tests/test_sale_order.py, (I hope i am not forgetting any step) :

- Create product_1
- create product_2
- create contract template 1 adding as order_line product_1
- create contract template 2 adding as order_line product_2
- set products as contract
- Create some orders of the products.
- Remove the orders buying product 1 and remove also the product 1 and template 1.
- Then error about "record doesn't exists" appears when trying to confirm a new quotation that includes product 2.

I had to remove the order line in contract template 2 and add it again to workaround the issue.
